Added additional boolean parameter "replace" to NavigationManager.NavigateTo

MariovanZeist opened the pull request:
Added optional boolean "replace" to Navigationmanager.NavigateTo

setting boolean to true will replace the uri in the browser history, instead of the default action that pushes the new uri onto the browsers history stack.

The core navigation infrastructure already has a functional "replace" parameter, but this parameter was not publicly exposed. 


Addresses #25540

---

## API review notes

This section is by @SteveSandersonMS and refers to the outcome I'm proposing with my suggestions below, rather than the current implementation.

The proposal is to add the following APIs in `Microsoft.AspNetCore.Components`, also showing selected existing APIs for context:

```diff
namespace Microsoft.AspNetCore.Components
{
    public abstract class NavigationManager
    {
        public void NavigateTo(string uri, bool forceLoad = false) { /* Calls NavigateToCore(string, bool) */ }
+       public void NavigateTo(string uri, NavigationOptions options) { /* Calls NavigateToCore(string, NavigationOptions) */ }

        protected abstract void NavigateToCore(string uri, bool forceLoad);
+       protected virtual void NavigateToCore(string uri, NavigationOptions options);
    }

+   [Flags]
+   public enum NavigationOptions
+   {
+       ForceLoad = 1,
+       ReplaceHistoryEntry = 2,
+   }
}
```

The main compromise in this design is around back-compatibility. I'm proposing we prioritise keeping people's existing unit test code working. If we didn't have to do that, we'd outright replace the existing `NavigateToCore` method with the new one, or at least obsolete the older one. However I don't see a way to obsolete the older one, since `NavigateTo(string, bool)` has to continue calling it, otherwise that will break anyone who has an existing mock or fake of NavigationManager.

I think the end result of this compromise is totally fine, in that for application developers, there's no drawback at all. The only people who have an extra method to overload is us (or other platform authors).

MariovanZeist committed:
Added additional boolean parameter "replace" to NavigationManager.NavigateTo
So instead of pushing the new uri into the browser history, we replace the uri in the browser history

SteveSandersonMS commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -102,7 +103,8 @@ public void NavigateTo(string uri, bool forceLoad = false)
         /// <param name="forceLoad">If true, bypasses client-side routing and forces the browser to load the new page from the server, whether or not the URI would normally be handled by the client-side router.</param>
-        protected abstract void NavigateToCore(string uri, bool forceLoad);
+        /// <param name="replace">If true, will replace the uri in the current browser history state, instead of pushing the new uri onto the browser history stack</param>
+        protected abstract void NavigateToCore(string uri, bool forceLoad, bool replace);
```
Thanks for the contribution, @MariovanZeist! This will be a great feature to have.

We will need to find a way to make this change without it being a breaking change. The way it's implemented here would mean that all existing subclasses of `NavigationManager` would become broken because they don't take a `replace` parameter.

One way this could be done would be like this:

```cs
// Leave this alone for back-compatibility
protected abstract void NavigateToCore(string uri, bool forceLoad);

protected virtual void NavigateToCore(string uri, bool forceLoad, bool replace)
{
    // Default implementation ignores "replace"
    NavigateToCore(uri, forceLoad);
}
```

Then, subclasses of `NavigationManager` can optionally override the 3-parameter overload of `NavigateToCore` and provide a more specialised implementation that respects `replace`. Other subclasses of `NavigationManager` will continue to work, but will just ignore the third parameter.

Another option would be that the 3-parameter overload of `NavigateToCore` should throw a `NotSupportedException` in the abstract base class rather than ignoring the third parameter. In this case, we'd want to have two separate overloads of `NavigateTo` - the existing one unchanged for back-compatibility, and a new three-parameter one that calls the three-parameter overload of `NavigateToCore`.

Finally, another design point to consider is making this more extensible in the future. What if we need a 4th parameter one day? One way to address this is defining some `struct NavigationOptions` that contains the `bool Replace` flag, and have that as the new third parameter. Then in the future we could add more flags or sub-parameters into that struct.

Does this make sense?

MariovanZeist replied:
Thanks @SteveSandersonMS for considering my PR, 
Yes I totally agree with your proposed changes, I wasn't considering binary compatibility but I should have when I created this PR.

I personally prefer the `struct NavigationOptions` version, just because of the extensibility. 

so my suggestion would something like :

```cs
//Add options class
public class NavigationOptions
{
    public bool ForceLoad { get; set;}
    public bool Replace { get; set; }
}

//Changes to NavigationManager

//keep this for backward compatibility
 public void NavigateTo(string uri, bool forceLoad = false)
{
  AssertInitialized();
  NavigateToCore(uri, new NavigationOptions { ForceLoad = forceLoad} );
}

//Add overload
public void NavigateTo(string uri, NavigationOptions options)
{
  AssertInitialized();
  NavigateToCore(uri, options);
}


//keep this for backward compatibility
protected abstract void NavigateToCore(string uri, bool forceLoad);

protected virtual void NavigateToCore(string uri, NavigationOptions options)
{
    // Default implementation ignores "replace"
    NavigateToCore(uri, options.ForceLoad);
}

```
Would these changes be acceptable?.

And does this break binary compatibility later if we decide to extend NavigationOptions ?

MariovanZeist committed:
Changes due to remarks

MariovanZeist commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -104,6 +116,19 @@ public void NavigateTo(string uri, bool forceLoad = false)
         /// <param name="forceLoad">If true, bypasses client-side routing and forces the browser to load the new page from the server, whether or not the URI would normally be handled by the client-side router.</param>
         protected abstract void NavigateToCore(string uri, bool forceLoad);
```
This does seem a bit odd to me,
We now have an abstract function which descendant classes are forced to implement,
But we would like them to override the function below to use the new NavigationOptions. 

A non breaking change would be to make this function virtual instead of abstract (which is non breaking) (And make it throw a NotImplemented exception)
And add an obsolete attribute to it, explaining that descendants should override NavigateToCore(string uri, NavigationOptions options) to implement routing.

This might be acceptable as typical users would not subclass NavigationManager,

SteveSandersonMS replied:
> A non breaking change would be to make this function virtual instead of abstract (which is non breaking) (And make it throw a NotImplemented exception)

I'm afraid I don't follow how the suggestion design would really work. Currently, `NavigationManager.NavigateTo` already calls `NavigateToCore(string, bool)`. Are you proposing we change `NavigateTo` to call the `(string, NavigationOptions)` overload instead?

 * If yes, then we need the `(string, NavigationOptions)` overload to have useful default behavior. And the only thing it could do is call through to the `(string, bool)` overload since that's what's implemented on existing subclasses. But then we need `(string, bool)` not to throw, since we're calling it ourselves.
 * If no, then we we're calling the `(string, bool)` overload ourselves and of course need it not to throw.

So I don't see how in either case we could have the `(string, bool)` overload throw.

> This might be acceptable as typical users would not subclass NavigationManager,

Perhaps you're suggesting we just have a breaking change. If that's the suggestion, we'd be better off just taking a source-breaking change and changing the `(string, bool)` overload to `(string, NavigationOptions)` without adding a second overload. But our goal here is not to have a breaking change.

MariovanZeist replied:
Sorry if I wasn't clear, 
 I am talking about the ``Navigationmanager.NavigateToCore`` function,

It used to be:
```cs
protected abstract void NavigateToCore(string uri, bool forceLoad);
````
But this now seems odd, to have an abstract function, that descendants need to override. when we are introducing a new NavigateToCore function with the new NavigationOptions

to make sure we don't introduce a breaking change I propose to change the abstract function to a virtual one.
And mark it as Obsolete

```cs
[Obsolete("Override NavigateToCore(string uri, NavigationOptions options) instead")]
protected virtual void NavigateToCore(string uri, bool forceLoad) => throw new NotImplementedException();

protected virtual void NavigateToCore(string uri, NavigationOptions options)
{
  //We call NavigateToCore(string uri, bool forceLoad) so to not introduce a breaking change.
  //derived classes should override this function to implement routing
  NavigateToCore(uri, options.ForceLoad);
}
```

This way, all users that use custom NavigationManager Descendants will continue to work. (Although I don't think there will be a lot of Custom NavigationManagers (Except maybe mocked ones) as it's not possible to register ones own NavigationManager ). But even then these descendants will still function as before. (If they override `` void NavigateToCore(string uri, bool forceLoad)`` As they have to because it was marked abstract,  their override will still be called.

And the Obsolete message will only be visible for people who create their own NavigationManager descendant, not the majority of users.

And just for completeness,
The Navigationmanager.NavigateTo original and overloads are:

```cs
public void NavigateTo(string uri, bool forceLoad = false)
{
  AssertInitialized();
  NavigateToCore(uri, new NavigationOptions { ForceLoad = forceLoad });
}

public void NavigateTo(string uri, NavigationOptions options)
{
  AssertInitialized();
  NavigateToCore(uri, options);
}
```

MariovanZeist replied:
Updated PR using https://github.com/dotnet/roslyn/blob/master/docs/Adding%20Optional%20Parameters%20in%20Public%20API.md

MariovanZeist committed:
Merge branch 'master' into MvZ_NavigationManagerReplace

zbecknell commented:
Thanks for working on this @MariovanZeist!

I want to bring up that there is a second part to #25540: I want to be able to replace the URL and not cause `OnParametersSet` to trigger.  Using the existing `replace` parameter on `navigateTo` will still cause `OnParametersSet` to fire, which probably WILL be the desired behavior most of the time.  

The good thing about now having `NavigationOptions` is that we can hopefully extend those options to include one for perhaps suppressing the `LocationChanged` event, which I think would achieve my goal.

So, for example, in `WebAssemblyNavigationManager`, `SetLocation()` could look like so:

```cs
public void SetLocation(string uri, bool isInterceptedLink, bool suppressLocationChanged)
{
    Uri = uri;

    if (suppressLocationChanged)
    {
        NotifyLocationChanged(isInterceptedLink);
    }
}
```

This would go towards achieving the following:

* Replace URL state
* Update `NavigationManager.Uri`
* _Don't_ trigger `OnParametersSet` for the current page

**Example usage:** representing the active tab in the URL: `mysite.com?activeTab=details`

@SteveSandersonMS Should I create a new ticket for the above purpose?  Does this sound like an acceptable solution?

SteveSandersonMS commented:
@zbecknell Can you give more details about why this will be a common requirement? If it’s a niche case we might recommend doing it via JS interop instead of having it in this API.

zbecknell commented:
Sure thing, @SteveSandersonMS!

We're coming up to about a year of developing a Blazor app for a new project with a small team, and it's amazing!  Thanks to this, we've encountered a good number of scenarios with a variety of challenges.  As an example scenario I have a page with the following route:

`/account/{accountId}`

When on this page, there is the possibility of navigating from say `/account/1` to `/account/2` via `NavigationManager` -- this causes `OnParametersSet` to trigger, but NOT `OnInitialized` since the page is already loaded, so we need to rely on `OnParametersSet` to know when to reload our page state for a different account.  

We want to assume that `OnParametersSet` triggering really means that a page parameter has changed, otherwise we'd need to keep track of each parameter's before and after state and manually allow the data to reload.

This is why I want to bypass the `LocationChanged` event, BUT I still need my `NavigationManager.Uri` to stay synced with the actual browser's URL, otherwise I'd need to use JsInterop to retrieve the actual URL, which necessarily needs to be async, which introduces yet more timing complexity. (I can't use the sync JsInterop because I dual-host client/server Blazor.)

I'm already using interop for my current solution, but it has a bug due to the `Uri` not being in sync.  I'm probably going to solve this in the interim by keeping track of my own `Uri` variable instead of relying on `NavigationManager`.

---

All that being said, this is why another proposal of mine was just letting me set the `Uri` for `NavigationManager` programmatically to manually keep it in sync, but I get why that would be likely to cause confusion, and thus isn't an acceptable solution.

SteveSandersonMS commented:
Thanks @zbecknell! That sounds worthy of a discussion in its own right. To keep this PR thread on topic, I've factored it out into a new issue at https://github.com/dotnet/aspnetcore/issues/25767. Hope that's OK!

MariovanZeist committed:
Merge remote-tracking branch 'upstream/master' into MvZ_NavigationManagerReplace

MariovanZeist committed:
Updated PublicApi
Modified NavigateTo be source and binary compatible with previous versions
See: to https://github.com/dotnet/roslyn/blob/master/docs/Adding%20Optional%20Parameters%20in%20Public%20API.md

MariovanZeist commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -90,10 +90,47 @@ protected set
+        /// (as returned by <see cref="BaseUri"/>).</param>
+        /// <param name="forceLoad">If true, bypasses client-side routing and forces the browser to load the new page from the server, whether or not the URI would normally be handled by the client-side router.</param>
+        /// <param name="replace">If true, will replace the uri in the current browser history state, instead of pushing the new uri onto the browser history stack.</param>
+        public void NavigateTo(string uri, bool forceLoad = false, bool replace = false)
```
Do we want this extra method as a convenience?, or force people who want to set the Replace flag to use the function
```cs
NavigateTo(string uri, NavigationOptions options)
``` 
Adding this function might introduce another backward compatibility problem later if we want to add even more options

MariovanZeist committed:
Called wrong Exception, fixed MVC test

SteveSandersonMS commented on src/Components/Components/src/NavigationOptions.cs:
```diff
@@ -0,0 +1,22 @@
+    /// </summary>
+    public class NavigationOptions
+    {
+        /// <summary>
+        ///If true, bypasses client-side routing and forces the browser to load the new page from the server, whether or not the URI would normally be handled by the client-side router.
+        /// </summary>
+        public bool ForceLoad { get; set; }
```
```suggestion
        public bool ForceLoad { get; init; }
```

SteveSandersonMS replied:
```suggestion
        /// <summary>
        /// Bypasses client-side routing and forces the browser to load the new page from the server, whether or not the URI would normally be handled by the client-side router.
        /// </summary>
        ForceLoad = 1,
```

SteveSandersonMS commented on src/Components/Components/src/NavigationOptions.cs:
```diff
@@ -0,0 +1,22 @@
+        /// </summary>
+        public bool ForceLoad { get; set; }
+        /// <summary>
+        /// If true, will replace the uri in the current browser history state, instead of pushing the new uri onto the browser history stack.
+        /// </summary>
+        public bool Replace { get; set; }
```
```suggestion
        public bool Replace { get; init; }
```

SteveSandersonMS replied:
```suggestion
        /// Indicates that the current history entry should be replaced, instead of pushing the new uri onto the history stack.
        /// </summary>
        Replace = 2,
```

SteveSandersonMS commented on src/Components/Components/src/NavigationOptions.cs:
```diff
@@ -0,0 +1,22 @@
+    /// <summary>
+    /// Additional options for navigating to another URI
+    /// </summary>
+    public class NavigationOptions
```
```suggestion
    public readonly struct NavigationOptions
```

SteveSandersonMS commented on src/Components/Components/test/Routing/RouterTest.cs:
```diff
@@ -181,8 +181,7 @@ internal class TestNavigationManager : NavigationManager
                 Initialize("https://www.example.com/subdir/", "https://www.example.com/subdir/jan");
-
-            protected override void NavigateToCore(string uri, bool forceLoad) => throw new NotImplementedException();
+            protected override void NavigateToCore(string uri, NavigationOptions options) => throw new NotImplementedException();
```
This change can and should be reverted, to show we're not relying on subclasses providing an implementation for the new overload.

SteveSandersonMS commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -90,10 +90,47 @@ protected set
+        /// <param name="uri">The destination URI. This can be absolute, or relative to the base URI
+        /// (as returned by <see cref="BaseUri"/>).</param>
+        /// <param name="options">Add additional navigation options <see cref="NavigationOptions"/>.</param>
+        public void NavigateTo(string uri, NavigationOptions options)
```
```suggestion
        public void NavigateTo(string uri, in NavigationOptions options)
```

Once it's a `readonly struct`, we can avoid copying by using `in`.

SteveSandersonMS commented on src/Components/WebAssembly/WebAssembly.Authentication/test/RemoteAuthenticationServiceTests.cs:
```diff
@@ -596,6 +596,6 @@ internal class TestNavigationManager : NavigationManager
             Initialize("https://www.example.com/base/", "https://www.example.com/base/add-product");
 
-        protected override void NavigateToCore(string uri, bool forceLoad) => throw new NotImplementedException();
+        protected override void NavigateToCore(string uri, NavigationOptions options) => throw new NotImplementedException();
```
Looks like the change here and the next few files can all be reverted with the above suggestions.

SteveSandersonMS reviewed (commented):
Thanks again for this, @MariovanZeist!

I hope you don't mind the large collection of suggestions I'm posting here. The core point I'm suggesting is that we can and should do this without it being a breaking change to people's existing unit tests. Everyone who has existing fake/mock implementations of `NavigationManager` should be able to keep their existing code working. We can achieve this by leaving the existing `NavigateToCore` alone, and having the new overload act as an independent parallel code path.

I know that makes it less of a clean change, and doesn't immediately get us to a world where people implementing `NavigationManager` only have a single `NavigateToCore` to override. However the only people who have to override both overloads will be framework/platform authors, and that's the sort of thing platform authors can figure out just fine. Application developers only subclass `NavigationManager` for unit tests, and in that case, they will only need to override the overload that their own unit test is actually using, so this will make sense.

Is it essential that, in the long term, we find some way to obsolete the `(string, bool)` overload? It's not clear to me that is essential. It more seems like a nice-to-have for framework developers, but won't make much (perhaps any) difference to application developers.

SteveSandersonMS commented:
@MariovanZeist Sorry for the disruption, but we had an API review meeting in which we realised that `NavigationOptions` should really be a `[Flags] enum` instead of `struct` or `class`.

Reasoning: a flags enum is sufficient for current use cases, extensible in the future if we end up with more flags, avoids the problems of both `struct` (can't be extended while retaining binary compatibility) and `class` (causes an allocation on every call), and if we really do need more complex options in the future, the flags enum can become part of whatever options class we end up using.

Hope this is OK!

MariovanZeist committed:
Synced with master

MariovanZeist committed:
Implemented API Review remarks.

!!!IMPORTANT!!!!

I have also added a pragma to disable a warning which should NOT be disabled.
It's only added to get it reviewed and fixed.

#pragma warning disable RS0027 // Public API with optional parameter(s) should have the most parameters amongst its public overloads.
Is added to the NavigationManager

Before merging this should be addressed.

MariovanZeist committed:
Reverted Encoding to version in master. No need for this file to change

MariovanZeist commented:
Sorry @javiercn and @pranavkm You were added because a file had their encoding changed. I reverted that change.

MariovanZeist reviewed (commented):
Applied remarks.
This makes the PR a lot cleaner and with fewer changes.
Making the NavigationOptions an Enum instead of a class is a better solution, especially as these are all just flags.

MariovanZeist reviewed (commented):
@SteveSandersonMS Thanks for the review. I have implemented your suggestions.
There is 1 issue that we need to resolve before we merge. That is the pragma I added.
Review comment is added to source code.

SteveSandersonMS commented on src/Components/Components/src/NavigationOptions.cs:
```diff
@@ -0,0 +1,27 @@
+        /// <summary>
+        /// Indicates that the current history entry should be replaced, instead of pushing the new uri onto the browser history stack.
+        /// </summary>
+        Replace = 2
```
Would it be OK to have a more explicit name for this? Otherwise it might be unclear what is being replaced:

```suggestion
        ReplaceHistoryEntry = 2
```

MariovanZeist replied:
I agree, the explicit name is better.

SteveSandersonMS commented on src/Components/Server/src/Circuits/RemoteNavigationManager.cs:
```diff
@@ -67,28 +67,33 @@ public void NotifyLocationChanged(string uri, bool intercepted)
             }
-
-            _jsRuntime.InvokeAsync<object>(Interop.NavigateTo, uri, forceLoad);
+            _jsRuntime.InvokeAsync<object>(Interop.NavigateTo, uri, (options & NavigationOptions.ForceLoad) !=0, (options & NavigationOptions.Replace) !=0);
```
Minor spacing tweak:

```suggestion
            _jsRuntime.InvokeAsync<object>(Interop.NavigateTo, uri, (options & NavigationOptions.ForceLoad) != 0, (options & NavigationOptions.Replace) != 0);
```

SteveSandersonMS commented on src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyNavigationManager.cs:
```diff
@@ -29,13 +29,18 @@ public void SetLocation(string uri, bool isInterceptedLink)
             }
-
-            DefaultWebAssemblyJSRuntime.Instance.Invoke<object>(Interop.NavigateTo, uri, forceLoad);
+            DefaultWebAssemblyJSRuntime.Instance.Invoke<object>(Interop.NavigateTo, uri, (options & NavigationOptions.ForceLoad) != 0, (options & NavigationOptions.Replace) != 0);
```
Line-length tweak:

```suggestion
            DefaultWebAssemblyJSRuntime.Instance.Invoke<object>(Interop.NavigateTo, uri,
                (options & NavigationOptions.ForceLoad) != 0,
                (options & NavigationOptions.Replace) != 0);
```

SteveSandersonMS commented on src/Components/Components/src/NavigationOptions.cs:
```diff
@@ -0,0 +1,27 @@
+        /// Use default options
+        /// </summary>
+        None = 0,
+        /// <summary>
+        ///bypasses client-side routing and forces the browser to load the new page from the server, whether or not the URI would normally be handled by the client-side router.
+        /// </summary>
+        ForceLoad =1,
```
```suggestion
        /// <summary>
        /// Bypasses client-side routing and forces the browser to load the new page from the server, whether or not the URI would normally be handled by the client-side router.
        /// </summary>
        ForceLoad = 1,
```

SteveSandersonMS committed:
Avoid RS0027 problem by replacing optional param with extra overload

SteveSandersonMS commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -90,12 +90,26 @@ protected set
         /// <param name="uri">The destination URI. This can be absolute, or relative to the base URI
         /// (as returned by <see cref="BaseUri"/>).</param>
         /// <param name="forceLoad">If true, bypasses client-side routing and forces the browser to load the new page from the server, whether or not the URI would normally be handled by the client-side router.</param>
+#pragma warning disable RS0027 // Public API with optional parameter(s) should have the most parameters amongst its public overloads.
```
Thanks for pointing this out. Yes, it's tricky!

I've added a commit to this PR which I think is the desired resolution: https://github.com/dotnet/aspnetcore/pull/25752/commits/6c12545067dae889ffe4e6e6bcbea1439c2ffa1e

The Roslyn analyzer is being pretty aggressive about how we keep track of binary compatibility, and according to its docs, the way to resolve this situation is to replace the old overload with optional params with two new overloads (details are at https://github.com/dotnet/roslyn/blob/master/docs/Adding%20Optional%20Parameters%20in%20Public%20API.md).

MariovanZeist replied:
Yeah, I wasn't sure if adding the overload you made would maintain Binary and Source compatibility. Thanks for the PR and explanation.

SteveSandersonMS commented:
This is great, @MariovanZeist. Thanks very much! I agree the current state of this PR is very good. It's a nice clean simple change.

The only thing missing before we can merge is test coverage. If you wanted, you could add unit test coverage for each of the `NavigateTo` overloads into the existing `NavigationManagerTest.cs`, showing that they translate into the expected `NavigateToCore` calls.

More importantly though, we need coverage in the E2E tests to show that the new APIs actually do what they claim to do in real browsers. In `src\Components\test\E2ETest\Tests\RoutingTest.cs`, you'll find existing E2E tests called `CanNavigateProgrammatically` and `CanNavigateProgrammaticallyWithForceLoad`. We'll need to add at least two further tests in a similar style:

 * One called `CanNavigateProgrammaticallyWithReplaceHistoryEntry`, which I guess should do a "replace" navigation, then trigger a "back", and verify that we end up on the page *before* the one that redirected, and that all of this was client-side navigation (see the existing test cases for how to verify that)
 * One called `CanNavigateProgrammaticallyWithForceLoadAndReplaceHistoryEntry`, which would be the same except we should verify that the initial navigation was a full-page load (see the existing test cases for how to verify that)

Is this possible? Thanks again!

MariovanZeist committed:
Implemented PR remarks.

MariovanZeist committed:
When navigateTo was called with both flags set (ForceLoad and Replace)
The forceLoad flag would be ignored if not loading the same Uri

MariovanZeist committed:
Added 3 e2e tests

MariovanZeist commented:
@SteveSandersonMS Besides the above-mentioned tests, I added a 3rd. This one checks if you do not specify ReplaceHistoryState, we indeed do not Replace the history state. (It also checks if the new overload NavigateTo(string uri) works as intended. This might be a little overkill, but better to have extra tests then no tests I reckon.

MariovanZeist commented:
@SteveSandersonMS During the testing of this PR I came across an issue that is related but not caused by this PR. (To be sure I am going to verify if the same problems occur on master) 
The issue is related to the ForceLoad flag and I see different behaviors across browsers. From working as intended, Reconnect issues, and downright crashing in js (getting the Reload option in the yellow bar)
I am testing Chrome / Edge and Firefox. 
This comment is just a quick heads up until I can tell you the issue in more detail.

SteveSandersonMS commented on src/Components/Web.JS/src/Services/NavigationManager.ts:
```diff
@@ -80,7 +80,11 @@ export function navigateTo(uri: string, forceLoad: boolean, replace: boolean = f
+      if (forceLoad) {
+        location.replace(uri);
+      } else {
+        history.replaceState(null, '', absoluteUri)
```
This logic looks wrong to me. What happens if `isWithinBaseUriSpace(absoluteUri)` is false, `force` is false, and `replace` is true?

What I'd expect is: since this is an external navigation, the force param is irrelevant, and the `replace` flag means we should be calling `location.replace`.

What the logic appears to do is call `history.replaceState`, which won't produce a useful behavior because it's only doing HTML5-style client-side navigation and won't trigger an actual external page load.

Here's the full truth table I'd expect:

| isWithinBaseUriSpace(absoluteUri) | Force | Replace | Expected behavior |
| ------------- |-------------| -----| ---| 
| false | false | false | `location.href = url` (append external page load to history) |
| false | false | true | `location.replace` (replace with external page load on history) |
| false | true | false | `location.href = url` (append external page load to history) |
| false | true | true | `location.replace` (replace with external page load on history) |
| true | false | false | `performInternalNavigation(absoluteUri, false, false)` (calls `history.pushState`) |
| true | false | true | `performInternalNavigation(absoluteUri, false, true)` (calls `history.replaceState`) |
| true | true | false | `location.href = url` (append external page load to history) |
| true | true | true | `location.replace` (replace with external page load on history) |

Do you agree with this list? Could you check how it corresponds to the implementation? I think the only combination that's currently wrong is the one I mentioned at the top of this comment.

I suspect that before this PR, the logic here was already wrong in that it didn't handle `replace` without also `force`. That was OK before because it wasn't really proper public API. But since this is becoming proper public API now, we should make sure all the combinations of parameters work correctly.

SteveSandersonMS commented on src/Components/test/E2ETest/Tests/RoutingTest.cs:
```diff
@@ -407,6 +407,122 @@ public void CanNavigateProgrammaticallyWithForceLoad()
+            Browser.True(() => Browser.Url.EndsWith("/TestHistory", StringComparison.Ordinal));
+            Browser.WaitUntilTestSelectorReady();
+        }
+
+
```
```suggestion

```

SteveSandersonMS commented:
@MariovanZeist This looks really good to me. Thanks for submitting it!

I think there's just [one more implementation issue to resolve](https://github.com/dotnet/aspnetcore/pull/25752#pullrequestreview-554711546) and then we should be ready to merge this.

SteveSandersonMS commented:
Continuing this work at https://github.com/dotnet/aspnetcore/pull/33751
